core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -105,16 +105,31 @@ class ReplicaFetcherThread(name:String,
       leaderEndOffset
     } else {
       /**
-       * The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
+       * If the leader's log end offset is greater than follower's log end offset, there are two possibilities:
+       * 1. The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
        * start offset because the leader has deleted old logs (log.logEndOffset < leaderStartOffset).
+       * 2. When leader migration occurs, there is a slight chance that the old leader's high watermark is greater than
+       * the new leader's log end offset, so an OffsetOutOfRangeException was thrown. After that some more messages are
+       * produced to the leader. When the follower was trying to handle the exception and query the log end offset of
+       * broker, the leader's log end offset become higher than follower's log end offset.
+       *
+       * In the first case, the follower's current log end offset is smaller than leader's log start offset. So the
+       * follower should truncate all its log, roll out a new segment and start to fetch from current leader's log
+       * start offset.
+       * In the second case, the follower should just keep the current log segments and retry fetch.
```

Comment:
Wait, we don't want messages on 90-100 to be different on two brokers. This can lead to horrible inconsistencies down the road.

The right thing it for broker 0 to find out that it needs to truncate to offset 90 (because that was the LEO of the new leader when broker 0 got demoted to follower) and fetch from broker 1 starting at offset 90.

If we can't make it happen, I think the current behavior (truncate to log start and fetch everything from there) is better - at least we preserve consistent view of the world between brokers.